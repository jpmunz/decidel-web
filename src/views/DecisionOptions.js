import React from "react";
import { connect } from "react-redux";
import { InputGroup, FormControl } from "react-bootstrap";
import {
  removeOption,
  addOption,
  undoOption,
  isCurrentTurn,
  getOptions,
  getCurrentUserName,
  getFinalDecision
} from "src/state";
import RemoveButton from "./RemoveButton";
import UndoButton from "./UndoButton";
import ImageButton from "./ImageButton";
import "./DecisionOptions.scss";
import wrench from "src/assets/wrench.svg";
import autoBind from "react-autobind";

export class DecisionOptions extends React.Component {
  constructor(props) {
    super(props);

    this.state = { newOption: "" };
    autoBind(this);
  }

  renderOption(option, i) {
    let action;
    let classNames = [];
    let numRemaining = this.props.options.filter(option => !option.isRemoved)
      .length;

    if (this.props.finalDecision === option.title) {
      classNames.push("final-decision");
    } else if (option.isRemoved) {
      classNames.push("removed");
    }

    if (!this.props.finalDecision) {
      if (option.undoId) {
        action = (
          <UndoButton
            onClick={() => this.handleUndoOptionClick(option.undoId)}
          />
        );
      } else if (this.props.isCurrentTurn && !option.isRemoved) {
        action = (
          <RemoveButton
            disabled={numRemaining === 1}
            title={
              numRemaining === 1 ? "You can't remove the last option!" : ""
            }
            onClick={() => this.handleRemoveOptionClick(i)}
          />
        );
      }
    }

    return (
      <li className={classNames.join(" ")} key={i}>
        <span className={"option-title"}>{option.title}</span>
        {action}
      </li>
    );
  }

  handleOptionKeyPress(event) {
    if (event.key === "Enter") {
      this.handleAddOptionClick();
    }
  }

  handleChangeNewOption(event) {
    this.setState({ newOption: event.target.value });
  }

  handleAddOptionClick() {
    this.props.addOption({
      id: this.props.id,
      title: this.state.newOption,
      userName: this.props.currentUserName
    });
    this.setState({ newOption: "" });
  }

  handleRemoveOptionClick(i) {
    this.props.removeOption({
      id: this.props.id,
      optionIndex: i,
      userName: this.props.currentUserName
    });
  }

  handleUndoOptionClick(undoId) {
    this.props.undoOption({
      id: this.props.id,
      undoId: undoId,
      userName: this.props.currentUserName
    });
  }

  render() {
    let addButton;

    if (this.props.isCurrentTurn && !this.props.finalDecision) {
      addButton = (
        <InputGroup className="add-option">
          <FormControl
            data-testid="add-option"
            type="text"
            onChange={this.handleChangeNewOption}
            onKeyPress={this.handleOptionKeyPress}
            placeholder="Toss a wrench in the plan by adding a new option"
            aria-label="newOption"
            value={this.state.newOption}
          />

          <InputGroup.Append>
            <ImageButton
              disabled={!this.state.newOption}
              onClick={this.handleAddOptionClick}
              alt="add"
              src={wrench}
            />
          </InputGroup.Append>
        </InputGroup>
      );
    }

    return (
      <div className="DecisionOptions">
        <ul>{this.props.options.map(this.renderOption)}</ul>
        {addButton}
      </div>
    );
  }
}

export default connect(
  (state, props) => {
    return {
      options: getOptions(state, props.id),
      finalDecision: getFinalDecision(state, props.id),
      currentUserName: getCurrentUserName(state),
      isCurrentTurn: isCurrentTurn(state, props.id)
    };
  },
  { removeOption, addOption, undoOption }
)(DecisionOptions);
