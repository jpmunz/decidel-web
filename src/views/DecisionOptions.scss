@import "mixins";
@import "variables";

.DecisionOptions {
  @include rounded();
  @include box-shadow(10px, 10px, 20px, rgba(0, 0, 0, 0.2));

  ul {
    padding: 5px 0;

    li {
      padding: 5px 10px;
      margin: 10px 5px;
      border: 1px solid #ddd;
      @include rounded();

      &.final-decision {
        border: 3px solid $black;

        .option-title {
          font-weight: bold;
        }
      }

      &.removed .option-title {
        text-decoration: line-through;
      }

      .ImageButton {
        float: right;
      }
    }
  }

  .add-option {
    margin: 0 5px;
    padding: 0 10px 20px;
    width: 100%;

    .form-control {
      border-top: none;
      border-left: none;
      border-right: none;
      border-radius: 0;
      padding: 0;
      box-shadow: none;
    }
  }
}
