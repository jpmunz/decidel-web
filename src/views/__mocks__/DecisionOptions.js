import React from "react";

export const MockDecisionOptions = jest.fn();

MockDecisionOptions.mockReturnValue(<div>Mocked DecisionOptions</div>);

const mock = jest.fn().mockImplementation(MockDecisionOptions);

export default mock;
