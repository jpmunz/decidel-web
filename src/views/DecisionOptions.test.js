import React from "react";
import { DecisionOptions } from "./DecisionOptions";
import { render, fireEvent } from "@testing-library/react";

const removeMock = jest.fn();
const addMock = jest.fn();
const undoMock = jest.fn();

const renderComponent = function({
  id = "22",
  options = [{ title: "option1" }, { title: "option2" }],
  finalDecision = false,
  isCurrentTurn = true,
  currentUserName = "Bob"
}) {
  return render(
    <DecisionOptions
      id={id}
      options={options}
      finalDecision={finalDecision}
      isCurrentTurn={isCurrentTurn}
      currentUserName={currentUserName}
      removeOption={removeMock}
      addOption={addMock}
      undoOption={undoMock}
    />
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("Another user's turn", () => renderTest({ isCurrentTurn: false }));
test("Decision is finalized", () =>
  renderTest({ finalDecision: "Picked Option" }));
test("Disallowing remove on the last option", () =>
  renderTest({ options: [{ title: "option1" }] }));

test("Adding options", () => {
  const component = renderComponent({});

  fireEvent.change(component.getByTestId("add-option"), {
    target: { value: "new option" }
  });
  fireEvent.click(component.getByAltText("add"));

  expect(addMock.mock.calls).toEqual([
    [
      {
        id: "22",
        title: "new option",
        userName: "Bob"
      }
    ]
  ]);
});

test("Removing options", () => {
  const component = renderComponent({
    options: [{ title: "option1" }, { title: "option2" }]
  });

  fireEvent.click(component.getAllByAltText("remove")[0]);
  expect(removeMock.mock.calls).toEqual([
    [
      {
        id: "22",
        optionIndex: 0,
        userName: "Bob"
      }
    ]
  ]);
});

test("Undoing changes", () => {
  const component = renderComponent({
    options: [
      { title: "option1" },
      { title: "option2", isRemoved: true, undoId: "a" },
      { title: "option3", undoId: "b" }
    ]
  });

  fireEvent.click(component.getAllByAltText("undo")[0]);
  fireEvent.click(component.getAllByAltText("undo")[1]);
  expect(undoMock.mock.calls).toEqual([
    [
      {
        id: "22",
        undoId: "a",
        userName: "Bob"
      }
    ],
    [
      {
        id: "22",
        undoId: "b",
        userName: "Bob"
      }
    ]
  ]);
});
