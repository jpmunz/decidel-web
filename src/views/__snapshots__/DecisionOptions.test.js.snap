// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Another user's turn 1`] = `
<DocumentFragment>
  <div
    class="DecisionOptions"
  >
    <ul>
      <li
        class=""
      >
        <span
          class="option-title"
        >
          option1
        </span>
      </li>
      <li
        class=""
      >
        <span
          class="option-title"
        >
          option2
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Decision is finalized 1`] = `
<DocumentFragment>
  <div
    class="DecisionOptions"
  >
    <ul>
      <li
        class=""
      >
        <span
          class="option-title"
        >
          option1
        </span>
      </li>
      <li
        class=""
      >
        <span
          class="option-title"
        >
          option2
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Disallowing remove on the last option 1`] = `
<DocumentFragment>
  <div
    class="DecisionOptions"
  >
    <ul>
      <li
        class=""
      >
        <span
          class="option-title"
        >
          option1
        </span>
        <button
          class="ImageButton btn btn-primary"
          disabled=""
          title="You can't remove the last option!"
          type="button"
        >
          <img
            alt="remove"
            src="remove.svg"
          />
        </button>
      </li>
    </ul>
    <div
      class="add-option input-group"
    >
      <input
        aria-label="newOption"
        class="form-control"
        data-testid="add-option"
        placeholder="Toss a wrench in the plan by adding a new option"
        type="text"
        value=""
      />
      <div
        class="remove-button-container input-group-append"
      >
        <button
          class="ImageButton btn btn-primary"
          disabled=""
          type="button"
        >
          <img
            alt="add"
            src="wrench.svg"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
